### PR TITLE
Fix up usages of Optionals

### PR DESCRIPTION
**Pull request by LiamClarkeNZ**
### Type of change

_Select the type of your PR_

- Refactoring

### Description

Fixed a few potentially troublesome patterns with Optionals and .get(), also fixed up some potential NPEs. 

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [X] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

**Comment by scholzj** [comment]
This seems to build on top of #5204 - It has the same issues. That makes it blocked by it. I think it will also cause rebase issues after #5204 is merged.

**Comment by LiamClarkeNZ** [comment]
> This seems to build on top of #5204 - It has the same issues. That makes it blocked by it. I think it will also cause rebase issues after #5204 is merged.

Oh cripes, the build failures definitely confirm that, I'll move it to draft until #5204 is finished and merged. Cheers @scholzj :)

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaBrokerConfigurationDiff.java [hunk 1]
```diff
@@ -90,10 +90,7 @@ private static void fillPlaceholderValue(Map<String, String> orderedProperties,
      */
     boolean isDesiredPropertyDefaultValue(String key, Config config) {
         Optional<ConfigEntry> entry = config.entries().stream().filter(configEntry -> configEntry.name().equals(key)).findFirst();
-        if (entry.isPresent()) {
-            return entry.get().isDefault();
-        }
-        return false;
+        return entry.map(ConfigEntry::isDefault).orElse(false);
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaBrokerConfigurationDiff.java:93]: Again here my preference would be simply 

```java
return config.entries().stream()
    .filter(configEntry -> configEntry.name().equals(key))
    .findFirst()
    .map(ConfigEntry::isDefault)
    .orElse(false);
```

**Review by tombentley — APPROVED** [review]
These changes LGTM. 

I left a couple of comments (essentially making the same point) about how we format stream computations with multiple stages. While it would be more churn to implement that everywhere in this PR I do think it would help readability and avoid some rather long lines. But better to get @scholzj's opinion before you proceed.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaClusterTest.java [hunk 2]
```diff
@@ -3263,26 +3263,26 @@ public void testGenerateDeploymentWithOAuthWithClientSecretAndTls() {
...
-        assertThat(sts.getSpec().getTemplate().getSpec().getVolumes().stream().filter(vol -> "oauth-plain-9092-0".equals(vol.getName())).findFirst().orElse(null).getSecret().getItems().get(0).getKey(), is("ca.crt"));
-        assertThat(sts.getSpec().getTemplate().getSpec().getVolumes().stream().filter(vol -> "oauth-plain-9092-0".equals(vol.getName())).findFirst().orElse(null).getSecret().getItems().get(0).getPath(), is("tls.crt"));
+        assertThat(sts.getSpec().getTemplate().getSpec().getVolumes().stream().filter(vol -> "oauth-plain-9092-0".equals(vol.getName())).findFirst().orElseThrow().getSecret().getItems().size(), is(1));
+        assertThat(sts.getSpec().getTemplate().getSpec().getVolumes().stream().filter(vol -> "oauth-plain-9092-0".equals(vol.getName())).findFirst().orElseThrow().getSecret().getItems().get(0).getKey(), is("ca.crt"));
+        assertThat(sts.getSpec().getTemplate().getSpec().getVolumes().stream().filter(vol -> "oauth-plain-9092-0".equals(vol.getName())).findFirst().orElseThrow().getSecret().getItems().get(0).getPath(), is("tls.crt"));
```
Comment (scholzj) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaClusterTest.java:3277]: Here and in many other places in the tests ... should we get the volume once, store it into a variable and then call the asserts on it? I know this was written badly already before this PR. But if you decide to improve it maybe it should be done properly and not just fix some IDE warnings?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java [hunk 3]
```diff
@@ -2027,17 +2027,13 @@ String nodePortAddressEnvVar(GenericKafkaListener listener)  {
...
                                             return null != node.getStatus().getAddresses().stream().filter(address -> hostIP.equals(address.getAddress())).findFirst().orElse(null);
                                         } else {
                                             return false;
                                         }
-                                    }).findFirst().orElse(null);
-
-                                    if (podNode != null) {
-                                        brokerNodes.put(podIndex, podNode);
-                                    }
+                                    }).findFirst().ifPresent(podNode -> brokerNodes.put(podIndex, podNode));
```
[the remark this answers is not in the document]
Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java:2036]: My view is that if it fits on one line it should be on one line. If it is on multiple lines, it is definitely better to have one stage per line.

[the remark this answers is not in the document]
Comment (LiamClarkeNZ) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java:2036]: :+1: Sweet as, will have a go at getting it at one stage per line.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaClusterTest.java (hunk 2 above)
Comment (LiamClarkeNZ) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaClusterTest.java:3277]: Thanks @scholzj, I'm trying to find the balance between specific changes only (in this case, the optional usages), and changes which lower the signal/noise ratio. I'll review my changes and add additional refactorings like the one you suggested, without (hopefully) going overboard :)

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaBrokerConfigurationDiff.java [hunk 4]
```diff
@@ -72,28 +72,11 @@
...
-     * @param key name of the property
-     * @param config Config which contains property
-     * @return true if property in desired map has a default value
-     */
-    boolean isDesiredPropertyDefaultValue(String key, Config config) {
-        Optional<ConfigEntry> entry = config.entries().stream().filter(configEntry -> configEntry.name().equals(key)).findFirst();
-        if (entry.isPresent()) {
-            return entry.get().isDefault();
-        }
-        return false;
```
Comment (LiamClarkeNZ) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaBrokerConfigurationDiff.java:96]: Removed this method as it was only used in unit test.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaBridgeClusterTest.java [hunk 5]
```diff
@@ -182,31 +190,44 @@ public void testGenerateService()   {
...
+        assertThat(container.getReadinessProbe().getInitialDelaySeconds(), is(healthDelay));
+        assertThat(container.getReadinessProbe().getTimeoutSeconds(), is(healthTimeout));
+
+        assertThat(container.getPorts().size(), is(1));
+
+        ContainerPort containerPort = container.getPorts().get(0);
+        assertThat(containerPort.getContainerPort(), is(KafkaBridgeCluster.DEFAULT_REST_API_PORT));
+        assertThat(containerPort.getName(), is(KafkaBridgeCluster.REST_API_PORT_NAME));
+        assertThat(containerPort.getProtocol(), is("TCP"));
+
```
Comment (LiamClarkeNZ) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaBridgeClusterTest.java:229]: Hi @scholzj, I wanted to check that this refactoring like many others in this file), pulling out commonly reused expressions that could be expressed as a single variable, doesn't go too far in your opinion. 

I've held back on further similar refactorings after this file, as I don't want to race ahead without your blessing.

Let me know :)

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaBridgeClusterTest.java [hunk 6]
```diff
@@ -417,8 +439,8 @@ public void testTemplate() {
                                 .withLabels(podLabels)
                                 .withAnnotations(podAnots)
                             .endMetadata()
-                            .withNewPriorityClassName("top-priority")
-                            .withNewSchedulerName("my-scheduler")
+                            .withPriorityClassName("top-priority")
+                            .withSchedulerName("my-scheduler")
```
Comment (LiamClarkeNZ) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaBridgeClusterTest.java:443]: withNew<X> was deprecated.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaBridgeClusterTest.java [hunk 7]
```diff
@@ -1058,4 +1095,8 @@ public void testCorsConfiguration() {
         assertThat(AbstractModel.containerEnvVars(container).get(KafkaBridgeCluster.ENV_VAR_KAFKA_BRIDGE_CORS_ALLOWED_ORIGINS), is("https://strimzi.io,https://cncf.io"));
         assertThat(AbstractModel.containerEnvVars(container).get(KafkaBridgeCluster.ENV_VAR_KAFKA_BRIDGE_CORS_ALLOWED_METHODS), is("GET,POST,PUT,DELETE,PATCH"));
     }
+
+    private EnvVar getEnvVar(List<EnvVar> envVars, String envVar) {
+        return envVars.stream().filter(var -> envVar.equals(var.getName())).findFirst().orElseThrow();
+    }
 }
```
Comment (LiamClarkeNZ) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaBridgeClusterTest.java:1102]: Extracted a very common pattern out to reduce scrolling sideways.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaBridgeClusterTest.java [hunk 8]
```diff
@@ -769,12 +803,14 @@ public void testGenerateDeploymentWithOAuthWithAccessToken() {
...
-        assertThat(cont.getEnv().stream().filter(var -> KafkaBridgeCluster.ENV_VAR_KAFKA_BRIDGE_OAUTH_ACCESS_TOKEN.equals(var.getName())).findFirst().orElse(null).getValueFrom().getSecretKeyRef().getKey(), is("my-token-key"));
-        assertThat(cont.getEnv().stream().filter(var -> KafkaBridgeCluster.ENV_VAR_KAFKA_BRIDGE_OAUTH_CONFIG.equals(var.getName())).findFirst().orElse(null).getValue().isEmpty(), is(true));
+        List<EnvVar> envVars = cont.getEnv();
+        assertThat(getEnvVar(envVars, KafkaBridgeCluster.ENV_VAR_KAFKA_BRIDGE_SASL_MECHANISM).getValue(), is("oauth"));
+        assertThat(getEnvVar(envVars, KafkaBridgeCluster.ENV_VAR_KAFKA_BRIDGE_OAUTH_ACCESS_TOKEN).getValueFrom().getSecretKeyRef().getName(), is("my-token-secret"));
+        assertThat(getEnvVar(envVars, KafkaBridgeCluster.ENV_VAR_KAFKA_BRIDGE_OAUTH_ACCESS_TOKEN).getValueFrom().getSecretKeyRef().getKey(), is("my-token-key"));
+        assertThat(getEnvVar(envVars, KafkaBridgeCluster.ENV_VAR_KAFKA_BRIDGE_OAUTH_CONFIG).getValue().isEmpty(), is(true));
     }
 
+
```
Comment (scholzj) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaBridgeClusterTest.java:813]: Is this needed?

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/resource/KafkaBrokerConfigurationDiffTest.java [hunk 9]
```diff
@@ -83,20 +83,6 @@ private void assertConfig(KafkaBrokerConfigurationDiff kcd, ConfigEntry ce) {
...
-        assertThat(kcd.isDesiredPropertyDefaultValue("offset.metadata.max.bytes", getCurrentConfiguration(emptyList())), is(true));
-    }
-
-    @Test
-    public void testNonDefaultValue() {
-        List<ConfigEntry> ces = singletonList(new ConfigEntry("offset.metadata.max.bytes", "4097"));
-        KafkaBrokerConfigurationDiff kcd = new KafkaBrokerConfigurationDiff(Reconciliation.DUMMY_RECONCILIATION, getCurrentConfiguration(ces),
-                getDesiredConfiguration(emptyList()), kafkaVersion, brokerId);
-        assertThat(kcd.isDesiredPropertyDefaultValue("offset.metadata.max.bytes", getCurrentConfiguration(ces)), is(false));
-    }
```
Comment (scholzj) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/resource/KafkaBrokerConfigurationDiffTest.java:98]: Why are these tests not needed anymore?

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaClusterTest.java (hunk 2 above)
Comment (scholzj) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaClusterTest.java:3277]: Fair point. I should not complain about too big PR one day and ask you to add more stuff next day :-). Feel free to split it into another PRs / create issues to cover it.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaBridgeClusterTest.java (hunk 5 above)
Comment (scholzj) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaBridgeClusterTest.java:229]: I'm probably fine with this. Although I do not think this was necessarily needed. 

I think there are several aspects to handle:
* Code readability
* Performance / resource consumption (The UT/IT tests now run for 30-40 minutes, so saving time is nice)

My particular concern in this respect are mainly the asserts where we do `stream()` several times in a row and search for exactly the same thing just to assert several different parts. That seems rather wasteful. I guess the getters being called here again and again should be less wasteful (although to be honest I'm just guessing - I'm not really expert on this kind of stuff).

Comment (LiamClarkeNZ) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaBridgeClusterTest.java:229]: Ah yep, good point on the reusage of stream().

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaBrokerConfigurationDiff.java [hunk 10]
```diff
@@ -131,8 +114,8 @@ public int getDiffSize() {
         return diff.size();
     }
 
-    private static boolean isIgnorableProperty(String key) {
-        return IGNORABLE_PROPERTIES.matcher(key).matches();
+    private static boolean isNotIgnorableProperty(String key) {
```
[the remark this answers is not in the document]
Comment (LiamClarkeNZ) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaBrokerConfigurationDiff.java:117]: Okay, no worries :)

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/resource/KafkaBrokerConfigurationDiffTest.java (hunk 9 above)
Comment (LiamClarkeNZ) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/resource/KafkaBrokerConfigurationDiffTest.java:98]: They were testing a method that was only called by those two tests.

**Review by scholzj — APPROVED** [review]

**Pull request merged**